### PR TITLE
ENT-8294 Only run clean_when_off FR bundle on FR hubs

### PR DESCRIPTION
**Pull request by craigcomstock**
Ticket: ENT-7968
Changelog: title

### CODE REVIEW

**Comment by craigcomstock** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7427/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7427/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7427/

**Comment by craigcomstock** [comment]
Interesting. With this change we get

```
Failed to see /opt/cfengine/federation/cftransport directory removed
```

in FR_basic_setup test. I will investigate.

**Comment by vpodzime** [comment]
Looks promising. :) Is there maybe a problem with the config file (not) existing at some important point?

**Comment by craigcomstock** [comment]
I think this will work now. Problem was that `am_superhub` and `am_feeder` relied on `target_state` being `on`, which it wouldn't be when changing it to `off`. So instead I guarded with the config existing and target_state off OR config not existing. :) @cf-bottom jenkins please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7428/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7428/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7428/

**Comment by vpodzime** [comment]
> So instead I guarded with the config existing and target_state off OR config not existing.

Config not existing means it will be triggered on hubs that are not FR hubs at all, right?

**Comment by craigcomstock** [comment]
> > So instead I guarded with the config existing and target_state off OR config not existing.
> 
> Config not existing means it will be triggered on hubs that are not FR hubs at all, right?

Yep. I think this is akin to the mess of "negative knowledge" problems CFEngine runs into. I'll refactor. :+1:

**Comment by craigcomstock** [comment]
@cf-bottom jenkins please.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7431/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7431/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7431/

**Comment by craigcomstock** [comment]
@nickanderson @vpodzime the failures in the last CI run are either known failures (apt wrapper not found in fr-tests) or related to apache restart problems with packaging aka https://tracker.mender.io/browse/ENT-8293 which I will work on next. So I think if this looks good we should merge and cherry pick.

**Comment by vpodzime** [comment]
It's nice that it makes CI green, but this is still a problem:
```
# cf-agent -KI
    info: Executing 'no timeout' ... '/var/cfengine/bin/psql cfsettings -c "TRUNCATE TABLE remote_hubs"'
  notice: Q: "...e/bin/psql cfse": TRUNCATE TABLE
    info: Last 1 quoted lines were generated by promiser '/var/cfengine/bin/psql cfsettings -c "TRUNCATE TABLE remote_hubs"'
    info: Completed execution of '/var/cfengine/bin/psql cfsettings -c "TRUNCATE TABLE remote_hubs"'
    info: Executing 'no timeout' ... '/var/cfengine/bin/psql cfsettings -c "TRUNCATE TABLE federated_reporting_settings"'
  notice: Q: "...e/bin/psql cfse": TRUNCATE TABLE
    info: Last 1 quoted lines were generated by promiser '/var/cfengine/bin/psql cfsettings -c "TRUNCATE TABLE federated_reporting_settings"'
    info: Completed execution of '/var/cfengine/bin/psql cfsettings -c "TRUNCATE TABLE federated_reporting_settings"'
```
Doing this over and over on **all hubs** is not great. If there's nothing else to do, we should at least suppress the info messages and the output.

**Comment by craigcomstock** [comment]
> It's nice that it makes CI green, but this is still a problem:
> 
> ```
> # cf-agent -KI
>     info: Executing 'no timeout' ... '/var/cfengine/bin/psql cfsettings -c "TRUNCATE TABLE remote_hubs"'
>   notice: Q: "...e/bin/psql cfse": TRUNCATE TABLE
>     info: Last 1 quoted lines were generated by promiser '/var/cfengine/bin/psql cfsettings -c "TRUNCATE TABLE remote_hubs"'
>     info: Completed execution of '/var/cfengine/bin/psql cfsettings -c "TRUNCATE TABLE remote_hubs"'
>     info: Executing 'no timeout' ... '/var/cfengine/bin/psql cfsettings -c "TRUNCATE TABLE federated_reporting_settings"'
>   notice: Q: "...e/bin/psql cfse": TRUNCATE TABLE
>     info: Last 1 quoted lines were generated by promiser '/var/cfengine/bin/psql cfsettings -c "TRUNCATE TABLE federated_reporting_settings"'
>     info: Completed execution of '/var/cfengine/bin/psql cfsettings -c "TRUNCATE TABLE federated_reporting_settings"'
> ```
> 
> Doing this over and over on **all hubs** is not great. If there's nothing else to do, we should at least suppress the info messages and the output.

Agreed. I'll do some more work on this. :+1:

**Comment by craigcomstock** [comment]
@cf-bottom jenkins please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7438/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7438/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7438/

File: cfe_internal/enterprise/federation/federation.cf [hunk 1]
```diff
@@ -332,11 +332,15 @@ bundle agent clean_when_off
   classes:
     selinux_enabled.default:_stdlib_path_exists_semanage::
       "has_cftransport_fcontext" expression => returnszero("$(paths.semanage) fcontext -l | grep $(home)", "useshell");
+    "remote_hubs_table_empty" expression => returnszero(`[ $(const.dollar)($(sys.bindir)/psql cfsettings --quiet --tuples-only --command "SELECT COUNT(*) FROM remote_hubs") -eq "0"] && true || false`);
```
Comment (vpodzime) [inline cfe_internal/enterprise/federation/federation.cf:335]: The `&& true || false` part is not necessary, AFAICT.

Comment (craigcomstock) [reply cfe_internal/enterprise/federation/federation.cf:335]: was an artifact of my doing echos there instead of true false. :+1:

**Comment by craigcomstock** [comment]
@cf-bottom jenkins please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7445/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7445/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7445/

**Review by vpodzime — APPROVED** [review]
Looks good to me.

**Comment by craigcomstock** [comment]
bootstrap-tests and sequential-tests failures were due to cloud rate limiting https://tracker.mender.io/browse/ENT-6915, rebuilding them now.

**Comment by craigcomstock** [comment]
sequential-tests failure likely has to do with preinstall scripts on debian platforms not finding `ss` or `netstat` commands: https://github.com/cfengine/buildscripts/pull/924

**Comment by craigcomstock** [comment]
sequential-tests failure was unrelated, probably apache restart/kill related https://tracker.mender.io/browse/ENT-8209 or https://tracker.mender.io/browse/ENT-8293

**Comment by craigcomstock** [comment]
cherry picks
https://github.com/cfengine/masterfiles/pull/2209 (3.18)
https://github.com/cfengine/masterfiles/pull/2210 (3.15)

**Pull request merged**